util/winutil/gp/policylock_windows.go

```diff
@@ -0,0 +1,275 @@
+//
+// To avoid deadlocks when acquiring both machine and user locks, acquire the
+// user lock before the machine lock.
+type PolicyLock struct {
+	scope Scope
+	token windows.Token
+
+	closing chan struct{} // closing is closed when the Close method is called.
+
+	mu      sync.Mutex
+	handle  atomic.Uintptr
+	lockCnt atomic.Int32 // A non-zero LSB indicates that the lock can be acquired.
+}
+
+// policyLockHandle is the underlying lock handle returned by enterCriticalPolicySection.
+type policyLockHandle uintptr
+
+type policyLockResult struct {
+	handle policyLockHandle
+	err    error
+}
+
+var (
+	// ErrInvalidLockState is returned by (*PolicyLock).Lock if the lock has a zero value or has already been closed.
+	ErrInvalidLockState = errors.New("the lock has not been created or has already been closed")
+	machinePolicyLock   lazy.SyncValue[*PolicyLock]
+)
+
+// NewMachinePolicyLock creates a PolicyLock that allows to pause application of the
+// computer policy. To avoid deadlocks when acquiring both machine and user
+// locks, acquire the user lock before the machine lock.
+func NewMachinePolicyLock() *PolicyLock {
+	return machinePolicyLock.Get(func() *PolicyLock {
+		lock := &PolicyLock{scope: MachinePolicy, closing: make(chan struct{})}
+		lock.lockCnt.Store(1) // mark as initialized
+		return lock
+	})
+}
+
+// NewUserPolicyLock creates a PolicyLock that allows to pause application of the
+// user policy for the specified user. To avoid deadlocks when acquiring both
+// machine and user locks, acquire the user lock before the machine lock.
+//
+// The token indicates which user's policy should be locked for reading.
+// If specified, the token must have TOKEN_DUPLICATE access,
+// the specified user must be logged in interactively.
+// and the caller retains ownership of the token.
+//
+// Otherwise, a zero token value indicates the current user. It should not
+// be used by services or other applications running under system identities.
+func NewUserPolicyLock(token windows.Token) (*PolicyLock, error) {
+	cs := &PolicyLock{scope: UserPolicy, closing: make(chan struct{})}
+	if token != 0 {
+		err := windows.DuplicateHandle(
+			windows.CurrentProcess(),
+			windows.Handle(token),
+			windows.CurrentProcess(),
+			(*windows.Handle)(&cs.token),
+			windows.TOKEN_QUERY|windows.TOKEN_DUPLICATE|windows.TOKEN_IMPERSONATE,
+			false,
+			0)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cs.lockCnt.Store(1) // mark as initialized
+	return cs, nil
+}
+
+// Lock locks l.
+// It returns ErrNotInitialized if l has a zero value or has already been closed,
+// or an Errno if the underlying Group Policy lock cannot be acquired.
+//
+// As a special case, it fails with windows.ERROR_ACCESS_DENIED
+// if l is a user policy lock, and the corresponding user is not logged in
+// interactively at the time of the call.
+func (l *PolicyLock) Lock() error {
+	if l == nil || l.lockCnt.Add(2)&1 == 0 {
```

Comment:
Rather than doing atomic arithmetic to construct a lock, is it possible to use existing primitives to do stuff in a way that's got fewer subtle things that need to be considered? Like is this performance critical enough that atomics make a meaningful difference vs taking `l.mu` every time? I get that `sync.RWMutex` won't work here because you need to be able to detect the "no readers -> one reader" and "one reader -> no reader" edges to tell Windows to do stuff. Or will Windows allow taking the read lock multiple times and then this package doesn't need to be nearly as complicated?

The `l == nil` should have a return without unlock if you want that to be permitted -- `l.Unlock` will panic if `l == nil`.

Reply:
Windows allows taking the read lock multiple times, but it's relatively slow, somewhat fragile, uses RPC internally, and can be acquired only a limited number of times simultaneously by all processes on the system. Therefore, I wanted to limit the number of times we call the underlying API. Additionally, the Group Policy Service can hold its write counterpart for prolonged periods, so I wanted to avoid having a blocking `Close()` method and also make the `Lock()` operation cancellable.

But the performance on our side is really not that critical, so I updated the code to take `l.mu` every time. Thanks!